homework-submission/reactweek1/src/App.js

```diff
@@ -0,0 +1,123 @@
+import React, { Component } from 'react';
+import './App.css';
+
+
+class App extends Component {
+  constructor(props) {
+    super(props);
+    this.changeToDone = this.changeToDone.bind(this);
+    this.descriptionHandler = this.descriptionHandler.bind(this);
+    this.deadLineHandler = this.deadLineHandler.bind(this);
+    this.removeHandler = this.removeHandler.bind(this);
+    this.addHandler = this.addHandler.bind(this);
+    this.state = {
+      list: [
+        {
+          "id": 1,
+          "description": "Get out of bed",
+          "deadline": "2017-09-11",
+          "done": true
+        },
+        {
+          "id": 2,
+          "description": "Brush teeth",
+          "deadline": "2017-09-10",
+          "done": false
+        },
+        {
+          "id": 3,
+          "description": "Eat breakfast",
+          "deadline": "2017-09-09",
+          "done": false
+        },
+      ],
+      newDescription: '',
+      newDeadline: ''
+    }
+  }
+
+  changeToDone = (event) => {
+    const id = event.target.id;
+    const listCopy = this.state.list.slice(0);
+    const itemIndex = listCopy.findIndex(el => { return el.id === Number(id) });
+    listCopy[itemIndex].done = !listCopy[itemIndex].done;
+    this.setState({
+      list: listCopy
+    });
+  }
+  descriptionHandler = (newDescription) => {
+    const state = this.state;
+    state['newDescription'] = newDescription;
+    this.setState(state)
+  };
+
+  deadLineHandler = (newDeadLine) => {
+    const state = this.state;
+    state['newDeadline'] = newDeadLine;
+    this.setState(state)
+  };
+
+  addHandler = () => {
+    const state = this.state;
+    const newItem = {
+      'id': this.state.list.length + 1,
+      'description': this.state.newDescription,
+      'deadline': this.state.newDeadline,
+      'done': false
+    };
+    this.state.list.push(newItem);
+    // this.state.newDeadline = '';
+    // this.state.newDeadline = ''; commented out these two cus it faces some problem through the test but dont know what the better way is
```

Comment:
it's a minor mistake and it's very common :)
you are using `this.state.newDeadline` twice, i think the second one is supposed to be `this.state. newDescription `?

Here you are mutating the state again :)
You can avoid that by using [the spread operator](https://zendev.com/2018/05/09/understanding-spread-operator-in-javascript.html):
```js
this.setState({list: [...state.list, newItem]});
```
What do you think?

Reply:
I think that you are absolutely right, but just for my information what does '...' mean in this line before state.list?

Reply:
you are right, i modified it the way that it is not mutating the state. Would be good if you take a look at it and let me know how you think about it.